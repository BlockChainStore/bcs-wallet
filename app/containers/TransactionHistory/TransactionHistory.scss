@import '../../styles/variables';
.transactionInfo {
    background: $main-background;
    opacity: 0.87;
    display: flex;
    flex-direction: column;
    text-align: center;
    position: relative;
    height: 100%;
    .columnHeader,
    .headerSpacer {
        flex: 0 0 auto;
    }
    .transactions {
        flex: 1 1 auto;
        overflow-y: auto;
    }
}

.columnHeader {
    position: relative;
    font-size: 1.2em;
    font-weight: 200;
}

.headerSpacer {
    width: 84%;
    margin: 15px 8%;
    border-top: 3px solid $spacer-color;
}

.updateLoader {
    right: 0;
    left: auto;
    top: -5px;
}