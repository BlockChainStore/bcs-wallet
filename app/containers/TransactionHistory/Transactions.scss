@import '../../styles/variables';
@import url('https://fonts.googleapis.com/css?family=Open+Sans');
.noTransactions {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%)
}

.transactionList {
    margin: 0;
    padding: 0;
    margin: 20px;
}

.row {
    display: flex;
    align-items: stretch;
    list-style-type: none;
    font-weight: 200;
    border-bottom: 1px solid $spacer-color;
    padding-bottom: 10px;
    margin-bottom: 10px;
    .txid {
        flex: 0 0 auto;
        color: $lite-link-color;
    }
    .amounts {
        margin-right: 10px;
        flex: 1 1 auto;
        white-space: nowrap;
    }
}

.txid {
    text-align: left;
    font-family: 'Open Sans', sans-serif;
    font-size: .9em;
}

.amounts {
    .amount {
        display: block;
        text-align: right;
        color: $secondatry-color;
    }
}