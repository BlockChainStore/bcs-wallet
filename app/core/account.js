// @flow
import { wallet } from 'neon-js'

import { getStorage, setStorage } from './storage'

function getNEP6AddressData (matchesEncryptedWIF: boolean, address: string) {
  if (matchesEncryptedWIF) {
    return { address }
  } else {
    return {}
  }
}

export async function upgradeNEP6AddAddresses (encryptedWIF: string, wif: string) {
  const data = getStorage('userWallet')
  const loggedIntoAccount = new wallet.Account(wif)

  if (data && data.accounts) {
    const accounts = data.accounts.map((account, idx) => ({
      ...account,
      ...getNEP6AddressData(account.key === encryptedWIF, loggedIntoAccount.address)
    }))

    await setStorage('userWallet', { ...data, accounts })
  }
}
