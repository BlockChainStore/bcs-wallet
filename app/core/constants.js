// @flow
export const NEON_WALLET_RELEASE_LINK =
  'https://github.com/CityOfZion/neon-wallet/releases'

export const ASSETS = {
  NEO: 'NEO',
  GAS: 'GAS'
}

export const EXPLORERS = {
  NEO_TRACKER: 'Neotracker',
  NEO_SCAN: 'Neoscan',
  ANT_CHAIN: 'Antchain',
  NEO_VERSE: 'NEOverse'
}

export const ROUTES = {
  HOME: '',
  DASHBOARD: '/dashboard',
  CREATE_WALLET: '/create',
  ENCRYPT_KEY: '/encrypt-key',
  LOGIN_PRIVATE_KEY: '/login-private-key',
  LOGIN_LOCAL_STORAGE: '/login-local-storage',
  LOGIN_LEDGER_NANO_S: '/login-ledger-nano-s',
  LOGIN_NEP2: '/login-encrypted',
  LOGIN_TOKEN_SALE: '/login-token-sale',
  TOKEN_SALE: '/token-sale',
  SETTINGS: '/settings',
  DISPLAY_WALLET_KEYS: '/display-wallet-keys'
}

export const NOTIFICATION_LEVELS = {
  ERROR: 'error',
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning'
}

export const NOTIFICATION_POSITIONS = {
  TOP_CENTER: 'tc',
  TOP_RIGHT: 'tr',
  TOP_LEFT: 'tl',
  BOTTOM_CENTER: 'bc',
  BOTTOM_RIGHT: 'br',
  BOTTOM_LEFT: 'bl'
}

export const BIP44_PATH =
  '8000002C' + '80000378' + '80000000' + '00000000' + '00000000'

export const MODAL_TYPES = {
  SEND: 'SEND',
  RECEIVE: 'RECEIVE',
  CONFIRM: 'CONFIRM',
  TOKEN_INFO: 'TOKEN_INFO',
  TOKEN: 'TOKEN',
  ICO: 'ICO'
}

export const MAIN_NETWORK_ID = '1'
export const TEST_NETWORK_ID = '2'

// TestNet
export const TOKENS_TEST = {
  BCS : 'e81c837d37763afbf894d319c868a09f219f2be9'
}

// MainNet
export const TOKENS = {
  BCS : '546c5872a992b2754ef327154f4c119baabff65f'
}

export const DEFAULT_WALLET = {
  name: 'userWallet',
  version: '1.0',
  scrypt: {
    cost: 16384,
    blockSize: 8,
    parallel: 8,
    size: 64
  },
  accounts: [],
  extra: null
}

export const DEFAULT_CURRENCY_CODE = 'usd'

export const CURRENCIES = {
  aud: { symbol: '$' },
  brl: { symbol: 'R$' },
  cad: { symbol: '$' },
  chf: { symbol: 'Fr.' },
  clp: { symbol: '$' },
  cny: { symbol: '¥' },
  czk: { symbol: 'Kč' },
  dkk: { symbol: 'kr. ' },
  eur: { symbol: '€' },
  gbp: { symbol: '£' },
  hkd: { symbol: '$' },
  huf: { symbol: 'Ft ' },
  idr: { symbol: 'Rp ' },
  ils: { symbol: '₪' },
  inr: { symbol: '₹' },
  jpy: { symbol: '¥' },
  krw: { symbol: '₩' },
  mxn: { symbol: '$' },
  myr: { symbol: 'RM' },
  nok: { symbol: 'kr ' },
  nzd: { symbol: '$' },
  php: { symbol: '₱' },
  pkr: { symbol: '₨ ' },
  pln: { symbol: 'zł' },
  rub: { symbol: '₽' },
  sek: { symbol: 'kr ' },
  sgd: { symbol: 'S$' },
  thb: { symbol: '฿' },
  try: { symbol: '₺' },
  twd: { symbol: 'NT$' },
  usd: { symbol: '$' },
  zar: { symbol: 'R ' }
}
