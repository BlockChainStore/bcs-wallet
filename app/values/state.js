// @flow
export type ProgressState = 'INITIAL' | 'LOADING' | 'LOADED' | 'FAILED'

export const INITIAL: ProgressState = 'INITIAL'
export const LOADING: ProgressState = 'LOADING'
export const LOADED: ProgressState = 'LOADED'
export const FAILED: ProgressState = 'FAILED'
