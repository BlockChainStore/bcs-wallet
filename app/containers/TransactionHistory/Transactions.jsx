// @flow
import React from 'react'
import classNames from 'classnames'

import Transaction from '../../components/Blockchain/Transaction'
import { ASSETS } from '../../core/constants'
import { isZero } from '../../core/math'
import { formatBalance } from '../../core/formatters'

import styles from './Transactions.scss'

type Props = {
  className?: string,
  transactions: Array<Object>
}

export default class Transactions extends React.Component<Props> {
  render () {
    const { className, transactions } = this.props

    if (transactions.length === 0) {
      return <div className={classNames(styles.noTransactions, className)}>No transactions</div>
    }

    return (
      <ul id='transactionList' className={classNames(styles.transactionList, className)}>
        {transactions.map((tx) => (
          <li key={tx.txid} className={styles.row}>
            <Transaction className={classNames(styles.txid, 'txid')} txid={tx.txid} />
            {this.renderAmounts(tx)}
          </li>
        ))}
      </ul>
    )
  }

  renderAmounts (tx: Object) {
    const forceRenderNEO = !isZero(tx[ASSETS.NEO]) || isZero(tx[ASSETS.GAS])

    return (
      <div className={styles.amounts}>
        {this.renderAmount(tx, ASSETS.NEO, forceRenderNEO)}
        {this.renderAmount(tx, ASSETS.GAS)}
      </div>
    )
  }

  renderAmount (tx: Object, symbol: SymbolType, forceRender: boolean = false) {
    const amount = tx[symbol]

    if (forceRender || !isZero(amount)) {
      return (
        <div className={classNames(styles.amount, `amount${symbol}`)}>
          {formatBalance(symbol, amount)} {symbol}
        </div>
      )
    }
  }
}
