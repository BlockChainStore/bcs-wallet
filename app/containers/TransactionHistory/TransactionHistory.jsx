// @flow
import React, { Component } from 'react'

import Loader from '../../components/Loader'

import Transactions from './Transactions'

import styles from './TransactionHistory.scss'

type Props = {
  transactions: Array<Object>,
  loading: boolean
}

export default class TransactionHistory extends Component<Props> {
  render () {
    const { transactions, loading } = this.props

    return (
      <div id='transactionInfo' className={styles.transactionInfo}>
        <div id='columnHeader' className={styles.columnHeader}>
          Transaction History {loading && <Loader className={styles.updateLoader} />}
        </div>
        <div className={styles.headerSpacer} />
        {!loading && (
          <Transactions className={styles.transactions} transactions={transactions} />
        )}
      </div>
    )
  }
}
