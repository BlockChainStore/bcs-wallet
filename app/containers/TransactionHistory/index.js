// @flow
import { compose, withProps } from 'recompose'

import TransactionHistory from './TransactionHistory'
import transactionHistoryActions from '../../actions/transactionHistoryActions'
import withFetch from '../../hocs/api/withFetch'
import withData from '../../hocs/api/withData'
import withProgressProp from '../../hocs/api/withProgressProp'
import withNetworkData from '../../hocs/withNetworkData'
import withAuthData from '../../hocs/withAuthData'
import withoutProps from '../../hocs/withoutProps'
import { LOADING } from '../../values/state'

const PROGRESS_PROP = 'progress'

const mapTransactionsDataToProps = (transactions) => ({
  transactions
})

const mapLoadingProp = (props) => ({
  loading: props[PROGRESS_PROP] === LOADING
})

export default compose(
  withNetworkData(),
  withAuthData(),
  withFetch(transactionHistoryActions),
  withData(transactionHistoryActions, mapTransactionsDataToProps),

  // pass `loading` boolean to component
  withProgressProp(transactionHistoryActions, { propName: PROGRESS_PROP }),
  withProps(mapLoadingProp),
  withoutProps(PROGRESS_PROP)
)(TransactionHistory)
