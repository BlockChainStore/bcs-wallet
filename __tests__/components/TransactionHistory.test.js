import React from 'react'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import { shallow, mount } from 'enzyme'
import { merge } from 'lodash'

import TransactionHistory from '../../app/containers/TransactionHistory'
import { MAIN_NETWORK_ID, EXPLORERS } from '../../app/core/constants'
import { LOADED } from '../../app/values/state'

const initialState = {
  api: {
    AUTH: {
      address: 'AWy7RNBVr9vDadRMK9p7i7Z1tL7GrLAxoh',
      wif: 'L4SLRcPgqNMAMwM3nFSxnh36f1v5omjPg3Ewy1tg2PnEon8AcHou'
    },
    NETWORK: {
      batch: false,
      state: LOADED,
      data: MAIN_NETWORK_ID
    },
    SETTINGS: {
      batch: false,
      state: LOADED,
      data: {
        blockExplorer: EXPLORERS.NEO_TRACKER
      }
    },
    TRANSACTION_HISTORY: {
      batch: false,
      state: LOADED,
      data: []
    }
  }
}

const transactions = [
  {
    NEO: '50',
    GAS: '0.00000000',
    txid: '76938979'
  },
  {
    NEO: '0',
    GAS: '0.40000000',
    txid: '76938980'
  }
]

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore([thunk])(state)

  let wrapper
  if (shallowRender) {
    wrapper = shallow(<TransactionHistory store={store} />)
  } else {
    wrapper = mount(
      <Provider store={store}>
        <TransactionHistory />
      </Provider>
    )
  }

  return {
    store,
    wrapper
  }
}

describe('TransactionHistory', () => {
  test('renders without crashing', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  test('correctly renders no transaction history', () => {
    const { wrapper } = setup(initialState, false)

    const columnHeader = wrapper.find('#columnHeader')
    expect(columnHeader.text()).toEqual('Transaction History ')

    const transactionList = wrapper.find('#transactionList')
    expect(transactionList.children().length).toEqual(0)
  })

  test('correctly renders with NEO and GAS transaction history', () => {
    const transactionState = merge({}, initialState, {
      api: { TRANSACTION_HISTORY: { data: transactions } }
    })
    const { wrapper } = setup(transactionState, false)

    const transactionList = wrapper.find('#transactionList')
    expect(transactionList.children().length).toEqual(2)
    expect(transactionList.childAt(0).find('.txid').first().text()).toEqual(transactions[0].txid)
    expect(transactionList.childAt(1).find('.txid').first().text()).toEqual(transactions[1].txid)
    expect(transactionList.childAt(0).find('.amountNEO').text()).toEqual('50 NEO')
    expect(transactionList.childAt(1).find('.amountGAS').text()).toEqual('0.40000000 GAS')
  })
})
