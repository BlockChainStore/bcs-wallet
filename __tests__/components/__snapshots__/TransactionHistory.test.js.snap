// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionHistory renders without crashing 1`] = `
<withData(withProps(Connect(withData(Connect(withFetch(Connect(withData(Connect(withProgressProp(withProps(withoutProps(TransactionHistory))))))))))))
  dispatch={[Function]}
  networkId="1"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
